### PR TITLE
Don't include declarations in usages (fixes #629).

### LINKED ISSUE #629: Go To -> Declaration or Usages from declaration should not include the declaration
This behavior is for version 1.20, which is currently the latest.

## Summary

A rule declaration is included in the results when selecting `Go To -> Declaration or Usages` from the context-sensitive menu of a rule declaration.  This behavior prevents IntelliJ from navigating from a declaration to a usage when there is only one usage.

## Steps to reproduce

1. Find or create a rule with only a single usage.
2. Select `Go To -> Declaration or Usages` from the context-sensitive menu of the declaration.

Screenshot 2023-04-16 at 8 53 58 AM

## Actual behavior

A popup containing two search results is displayed. The two search results are the declaration and the usage.

Screenshot 2023-04-16 at 8 54 14 AM

## Expected behavior

Contrast this behavior with the behavior when performing the same steps on Java code.

Screenshot 2023-04-16 at 8 31 05 AM

The behavior on Java code is that IntelliJ navigates from the declaration to the usage.

Screenshot 2023-04-16 at 8 31 22 AM

## Intended behavior?

The heading "Expected  behavior" is referring to the behavior that I expected.  I wonder though if the actual behavior is the behavior intended by the ANTLR plugin maintainers.  In the `jetbrains-plugin-sample` repository, [this javadoc comment](https://github.com/antlr/jetbrains-plugin-sample/blob/0b8d515b5d38e4240d79d69a5e00c7ead0bf301a/src/main/java/org/antlr/jetbrains/sample/psi/IdentifierPSINode.java#L28-L29) says

> we want defs to be included in "find usages."

I don't think definitions/declarations should be included in the list of usages, but I can understand if others think they should be included.

### PR DESCRIPTION
**Pull request by ris58h**

### CODE REVIEW

**Comment by ris58h** [comment]
@bjansen Sorry for bothering you. I don't know who should I ping to approve the workflow.

**Comment by bjansen** [comment]
I'll look into it this week, sorry for the delay

File: src/main/java/org/antlr/intellij/plugin/ANTLRv4ASTFactory.java [hunk 1]
```diff
@@ -29,6 +29,7 @@ public class ANTLRv4ASTFactory extends ASTFactory {
 		ruleElementTypeToPsiFactory.put(ANTLRv4TokenTypes.RULE_ELEMENT_TYPES.get(ANTLRv4Parser.RULE_modeSpec), ModeSpecNode.Factory.INSTANCE);
 		ruleElementTypeToPsiFactory.put(ANTLRv4TokenTypes.RULE_ELEMENT_TYPES.get(ANTLRv4Parser.RULE_action), AtAction.Factory.INSTANCE);
 		ruleElementTypeToPsiFactory.put(ANTLRv4TokenTypes.RULE_ELEMENT_TYPES.get(ANTLRv4Parser.RULE_identifier), TokenSpecNode.Factory.INSTANCE);
+		ruleElementTypeToPsiFactory.put(ANTLRv4TokenTypes.RULE_ELEMENT_TYPES.get(ANTLRv4Parser.RULE_tokensSpec), TokensSpecNode.Factory.INSTANCE);
```
Comment (bjansen) [inline src/main/java/org/antlr/intellij/plugin/ANTLRv4ASTFactory.java:32]: I'm not sure we need this, token definitions already have a parent of type `TokenSpecNode`:
image
So using `getParent() instanceof TokenSpecNode` should be enough.

File: src/main/java/org/antlr/intellij/plugin/psi/LexerRuleRefNode.java [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
 
 public class LexerRuleRefNode extends GrammarElementRefNode {
 	public LexerRuleRefNode(IElementType type, CharSequence text) {
 		super(type, text);
 	}
 
 	@Override
-	public PsiReference getReference() {
-		return new GrammarElementRef(this, getText());
+	protected boolean isDeclaration() {
```
Comment (bjansen) [inline src/main/java/org/antlr/intellij/plugin/psi/LexerRuleRefNode.java:12]: I would have kept `getReference()` here and simply add this snippet:
```
		if (getParent() instanceof LexerRuleSpecNode || getParent() instanceof TokenSpecNode) {
			// A declaration doesn't need to reference itself
			return null;
		}
```

(same in `ParserRuleRefNode`)

**Review by bjansen — CHANGES_REQUESTED** [review]

**Comment by ris58h** [comment]
@bjansen I've done the requested changes.

**Comment by bjansen** [comment]
Thanks for the contribution!

**Pull request merged**